### PR TITLE
Add guard statements

### PR DESCRIPTION
**Pull request by jeongm-in**
# Overview
First of all, title of PR is a bit misleading. 
In the original code, variables `currentTab.url` and `currentTab.title` had type of `string | undefined`.
So I included a [type guard statements](https://github.com/jeongm-in/webmark/blob/41c6fbd7a7c892b422ee1d39c7afb99ea7c43b2a/src/index.tsx#L111), in an attempt to guarantee the type of `thisTabUrl` and `thisTabTitle` to be `string`. 
In any case where url and title are type `undefined`, these will resolve to "https://example.com/error" and "read later" respectively. 
These "sanitized" url and title string variables are then passed onto other functions for bookmark registration. 
@Bartleby2718 suggested setting url and title as optional variables in functions that use these variables. 

# Changes
As a result
- `saveIfNotAlreadyThere` now takes parameters `webmarkFolderId: string, url?: string, title?: string`.
- `isInTree` now takes parameters `nodes: chrome.bookmarks.BookmarkTreeNode[], url?: string`.
  - `isInTree` originally took parameters `url: string, nodes: chrome.bookmarks.BookmarkTreeNode[]`, but I switched orders because optional parameters cannot precede required parameters.
- `saveWithConfidence` now takes parameters `webmarkFolderId: string, url?: string, title?: string`.

In addition to function parameters, `thisTabUrl` and `thisTabTitle` now have type `string | undefined`. 

# Test
After changing these parameters and variable types, I was afraid what would happen if any of the url or title strings are `undefined`. 

I created sample sites so that I could test the updated extension. 

title-test 

1. [Empty Title Tag](https://jeongm.in/~jeongmin/emptyTitle.html)
  - `<title></title>`
  - Extension saved this website's url as its title.
2. [No Title Tag](https://jeongm.in/~jeongmin/noTitle.html)
  - Extension saved this website's url as its title.
3. [Title Tag](https://jeongm.in/~jeongmin/normalPage.html)
  - `<title>Title</title>`
  - Extension correctly saved this website's title as its title.
 
I also tested what would happen with sites with empty url. 
- Chrome new tabs page, and it correctly saved new tabs page as bookmark. However, new tab page actually has an url of "chrome://newtab".

# Conclusion
- In conclusion, websites with empty or no title tags have their url as titles.  
- I was unable to find how to access a webpage without an url. 
- Another approach to this would be using non null assertion operator `!` after variables `thisTabUrl` and `thisTabTitle` to "force" them to be `string` type.

### CODE REVIEW

**Review by Bartleby2718 — COMMENTED** [review]
1. I like how you tested, with your own website, what happens when there is no title tag. That's something I can't do.
2. Because of your tests, I think we can now rest assured that in the callback of [`chrome.tabs.query `](https://developer.chrome.com/extensions/tabs#method-query), [Tab](https://developer.chrome.com/extensions/tabs#type-Tab) will always have a `title` (because our `manifest.json` includes the "tabs" permission). That is, `const thisTabUrl: string = currentTab.url!;`.
3. I don't really understand why `thisTabUrl` and `thisTabTitle` are still `string | undefined`. I did say:
> Also, using ? may be better than declaring it string | undefined. 

However, I think the current way of sanitizing with a ternary operator in `getCurrentUrlAndSave ` is way better than using `?`. (i.e. ternary operator(`a?b:c`) >> existential operator(`parameter?`) > `: string | undefined`) If `url` and `title` become optional parameters, the uncertainty propagates to other functions like `isInTree`, `saveIfNotAlreadyThere`, and `saveWithConfidence`.

**Comment by Bartleby2718** [comment]
Since the external behavior stays the same, I'll change the label.
> Refactoring is a disciplined technique for restructuring an existing body of code, altering its internal structure without changing its external behavior.

https://www.refactoring.com/ (Martin Fowler's website)

**Comment by jeongm-in** [comment]
Well see for yourself with commit [5149059](https://github.com/jeongm-in/webmark/pull/61/commits/514905957435ce6eac1c015fd63f202a4b1cbcd8). If I leave type of those two variables as `string`, it will throw error at me because return value of `.url` and `.title` are `string | undefined`.

**Comment by jeongm-in** [comment]
But if we use ternary operator `a ? b : c`, what goes in `b`?

**Comment by Bartleby2718** [comment]
a) If you read 2. carefully, I added [an exclamation work(`!`)](https://stackoverflow.com/questions/42273853/in-typescript-what-is-the-exclamation-mark-bang-operator-when-dereferenci) at the end.
b) Well, I guess that's why we should go with the exclamation mark then.
